lib/Sema/MiscDiagnostics.cpp

```diff
@@ -2525,9 +2525,109 @@ class VarDeclUsageChecker : public ASTWalker {
   }
 
 };
+  
+/// An AST walker that determines the underlying type of an opaque return decl
+/// from its associated function body.
+class OpaqueUnderlyingTypeChecker : public ASTWalker {
+  TypeChecker &TC;
+  AbstractFunctionDecl *Implementation;
+  OpaqueTypeDecl *OpaqueDecl;
+  SmallVector<std::pair<Expr*, Type>, 4> Candidates;
+public:
+  OpaqueUnderlyingTypeChecker(TypeChecker &TC,
+                              AbstractFunctionDecl *Implementation,
+                              OpaqueTypeDecl *OpaqueDecl)
+    : TC(TC),
+      Implementation(Implementation),
+      OpaqueDecl(OpaqueDecl)
+  {
+    
+  }
+  
+  void check() {
+    Implementation->getBody()->walk(*this);
+    
+    // If there are no candidates, then the body has no return statements, and
+    // we have nothing to infer the underlying type from.
+    if (Candidates.empty()) {
+      TC.diagnose(Implementation->getLoc(),
+                  diag::opaque_type_no_underlying_type_candidates);
+      return;
+    }
+    
+    // Check whether all of the underlying type candidates match up.
+    auto opaqueTypeInContext =
+      Implementation->mapTypeIntoContext(OpaqueDecl->getDeclaredInterfaceType());
+    Type underlyingType = Candidates.front().second;
+    
+    bool mismatch = false;
+    for (auto otherCandidate : llvm::makeArrayRef(Candidates).slice(1)) {
+      // Disregard tautological candidates.
+      if (otherCandidate.second->isEqual(opaqueTypeInContext))
+        continue;
+        
+      if (!underlyingType->isEqual(otherCandidate.second)) {
```

Comment:
If you really wanted to be fancy, use TypeJoin here

Reply:
I was going to leave it up to evolution to decide whether we should try to join divergent return types. I can see the motivation, but it also seems like it could cover up mistakes just as easily.